### PR TITLE
Feature/undo

### PR DESCRIPTION
**Pull request by cortfritz**
Provides a subset of yrs & yjs UndoManager for Yex.Text, Yex.Map, and Yex.Array - no XML yet.

Has:
- undo
- redo
- multiple scopes (no scope removal yet)
- multiple origins and exclude origin
- stop tracking
- clear
- set merge batch timeout with options
- observer genserver (imho should be generalized - for now is undo observer specifically)
- stack item metadata

## Summary by CodeRabbit

- **New Features**
	- Introduced an Undo Manager for managing undo/redo operations with customizable options.
	- Added support for tracking origins and observers in the undo process.
	- Expanded functionality with new data types and event handling for undo operations.
	- Specified Elixir version for development environment management.
	- Enhanced undo management system with new functions for managing undo operations.
	- Added a new atom for message identification in the system.

- **Bug Fixes**
	- Improved error handling and state management during undo and redo operations.

- **Tests**
	- Comprehensive test suite for the Undo Manager, validating various functionalities and edge cases.
	- Added tests for new undo operations and edge cases related to document types.

### CODE REVIEW

**Comment by cortfritz** [comment]
@satoren I have pulled out the observer code, and updated to pass checks.

The problem with the observer is that:
1. the undo observer is designed to event on changes to the undo stack: added, popped, and updated.
2. it allows, perhaps among other purposes, for updating metadata in the undo stack
3. this is useful: e.g. saving cursor position for a specific change event, so then when we undo/redo we place the cursor in the correct place
4. looking at the yrs test of this (see below), changes to metadata are effected via meta_mut

I was able to get eventing to work so that elixir code could register callbacks, but I haven't yet figured out how to safely mutate the metadata in that elixir code across the NIF boundary.

There may be a non-blocking transactional strategy.  Or maybe we find a non-racey way to have a separate `update_meta` NIF that can target the correct item in the stack and take a metadata update.  I tried a bunch of things but none were great yet.  I'll keep going.

In the meantime, here is a working subset of UndoManager that handles everything in your yjs example linked in this thread, except for the eventing and metadata.

Here is the relevant yrs test:
```
#[test]
    fn undo_events() {
        use crate::undo::UndoManager;
        type Metadata = HashMap<String, usize>;

        let doc = Doc::with_client_id(1);
        let txt = doc.get_or_insert_text("test");
        let mut mgr: UndoManager<Metadata> = UndoManager::new(&doc, &txt);

        let result = Arc::new(AtomicUsize::new(0));
        let counter = AtomicUsize::new(1);

        let txt_clone = txt.clone();
        let _sub1 = mgr.observe_item_added(move |_, e| {
            assert!(e.has_changed(&txt_clone));
            let c = counter.fetch_add(1, Ordering::SeqCst);
            let e = e.meta_mut().entry("test".to_string()).or_default();
            *e = c;
        });

        let txt_clone = txt.clone();
        let result_clone = result.clone();
        let _sub2 = mgr.observe_item_popped(move |_, e| {
            assert!(e.has_changed(&txt_clone));
            if let Some(&v) = e.meta_mut().get("test") {
                result_clone.store(v, Ordering::Relaxed);
            }
        });

        txt.insert(&mut doc.transact_mut(), 0, "abc");
        mgr.undo_blocking();
        assert_eq!(result.load(Ordering::SeqCst), 1);
        mgr.redo_blocking();
        assert_eq!(result.load(Ordering::SeqCst), 2);
    }
    ```

File: lib/undo_manager.ex [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+  @doc false
+  def unwrap_manager_result({:ok, manager}), do: manager
+  def unwrap_manager_result(error), do: error
+
+  @doc """
+  Creates a new UndoManager for the given document and scope with default options.
+  The scope can be a Text, Array, or Map type.
+  """
+  def new(doc, %Yex.Text{} = scope) do
```
Comment (satoren) [inline lib/undo_manager.ex:35]: https://github.com/satoren/y_ex/blob/main/lib/shared_type/shared_type.ex
https://github.com/satoren/y_ex/blob/d9b3d71d76fb3070e8f436fdcb86107d0ca06de3/native/yex/src/yinput.rs#L333

If NifUntaggedEnum is used, tag is not required.

Comment (cortfritz) [reply lib/undo_manager.ex]: I removed the unnecessary matching on Yex.Text, Yex.Array, and Yex.Map and just pass `scope` and this works.  But left the SharedType struct as NifTaggedEnum.  Changing it to NifUntaggedEnum would require other changes.  Is that ok?

Comment (cortfritz) [reply lib/undo_manager.ex]: since I'm using your NifSharedTypeEnum I added tests for XML types also.

**Comment by cortfritz** [comment]
@satoren - i'm going to address the coderabbit feedback.

I'd like to confirm with you one style change.  See the first coderabbit feedback regarding unwrapping`new` and `new_with_options`.  

## why i did what i did
Following the style of Yex.Doc.new() returning a doc without a {:ok, doc} tuple,  I have UndoManager.new(doc, scope) and UndoManager.new_with_options(doc, scope, options) returning a naked UndoManager instead of {:ok, undo_manager} .  

## why i am accepting coderabbit feedback
Possibly my choice of returning a naked UndoManager is wrong because unlike Doc.new(), UndoManager.new(doc, scope) could error because inputs are invalid, so I should return an :ok, value tuple or error.

If you would prefer I return a naked UndoManager, let me know.

File: native/yex/src/atoms.rs [hunk 2]
```diff
@@ -34,4 +34,13 @@ rustler::atoms! {
...
   attributes,
+
+  // Type identifiers for branches
+  text,
+  map,
+  array,
+  xml_element,
+  xml_fragment,
+  xml_text,
+  popped,
```
Comment (satoren) [inline native/yex/src/atoms.rs:45]: Are these still in use?

**Comment by satoren** [comment]
@cortfritz Thank you for your excellent work.
Since the scope can become invalid, it seems correct, as you mentioned, to return an :ok tuple.

File: native/yex/src/atoms.rs (hunk 2 above)
Comment (cortfritz) [reply native/yex/src/atoms.rs:45]: i don't think so - commenting them out and compiling i still pass all tests.  Also it looks like shared types doesn't need atoms at all anymore.

File: lib/undo_manager.ex [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+      when is_struct(scope, Yex.Text) or
+             is_struct(scope, Yex.Array) or
+             is_struct(scope, Yex.Map) or
+             is_struct(scope, Yex.XmlText) or
+             is_struct(scope, Yex.XmlElement) or
+             is_struct(scope, Yex.XmlFragment) do
+    if is_struct(scope) do
+      new_with_options(doc, scope, %Options{})
+    else
+      {:error, "Invalid scope: expected a struct"}
```
Comment (satoren) [inline lib/undo_manager.ex:47]: It doesn't seem to reach here because it is checked by when.

File: lib/undo_manager.ex [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+             is_struct(scope, Yex.Map) or
+             is_struct(scope, Yex.XmlText) or
+             is_struct(scope, Yex.XmlElement) or
+             is_struct(scope, Yex.XmlFragment) do
+    cond do
+      not is_struct(doc, Yex.Doc) ->
+        {:error, "Invalid document: expected a Yex.Doc struct"}
+
+      not is_struct(scope) ->
+        {:error, "Invalid scope: expected a struct"}
```
Comment (satoren) [inline lib/undo_manager.ex:78]: It doesn't seem to reach here because it is checked by when.

Comment (cortfritz) [reply lib/undo_manager.ex]: will fix, thank you very much.

**Comment by cortfritz** [comment]
@satoren there is a real problem with updating metadata across the NIF boundary.  I'd like to provide context here, and options, and get your preference for how to proceed.

metadata needs to be stored with items in the undo stack.  this allows for things like adding cursor position to a change so that when we undo, we can set the cursor position correctly.

the way yrs handles this is to observe undo add and pop and provide a mutable undo item to the callback.  so we can set metadata on `added` and get the metadata back on `popped` 

here is the yrs test that exercises this:

```
#[test]
    fn undo_events() {
        use crate::undo::UndoManager;
        type Metadata = HashMap<String, usize>;

        let doc = Doc::with_client_id(1);
        let txt = doc.get_or_insert_text("test");
        let mut mgr: UndoManager<Metadata> = UndoManager::new(&doc, &txt);

        let result = Arc::new(AtomicUsize::new(0));
        let counter = AtomicUsize::new(1);

        let txt_clone = txt.clone();
        let _sub1 = mgr.observe_item_added(move |_, e| {
            assert!(e.has_changed(&txt_clone));
            let c = counter.fetch_add(1, Ordering::SeqCst);
            let e = e.meta_mut().entry("test".to_string()).or_default();
            *e = c;
        });

        let txt_clone = txt.clone();
        let result_clone = result.clone();
        let _sub2 = mgr.observe_item_popped(move |_, e| {
            assert!(e.has_changed(&txt_clone));
            if let Some(&v) = e.meta_mut().get("test") {
                result_clone.store(v, Ordering::Relaxed);
            }
        });

        txt.insert(&mut doc.transact_mut(), 0, "abc");
        mgr.undo_blocking();
        assert_eq!(result.load(Ordering::SeqCst), 1);
        mgr.redo_blocking();
        assert_eq!(result.load(Ordering::SeqCst), 2);
    }

```

The problem is that we shouldn't allow the callback we pass in to update the mutable event item across the NIF boundary.

possible solutions:
1. create an asynchronous queue so the callback is non blocking and a worker process works on the queue setting the metadata.  yrs doesn't provide a "update event metadata" function so we will have to get the stack, search for the event item with the id, then change it.  complex!  slow! expensive!
2. do not set metadata on the rust-side event, but on the elixir side, store meta data in a matching store so that we keep metadata for each event item in an elixir-local store e.g. a genserver or ets.  the state would be indexed by undomanager reference (each undomanager instance or it's copy would have it's own store) and event item id (per event, within the scope of the undomanager's store)

both of these seem super clanky.  my preference is 2 for performance reasons.  it fails if your design principle is that Yex.UndoManager can be used at the same time someone is using the underlying yrs directly.  because that other use wouldn't have access to the metadata stored on the elixir side.


thoughts?

**Comment by satoren** [comment]
@cortfritz 
It seems useful even without the observe. Shall we go ahead and merge this PR for now, or would it be better to wait until everything is complete?

**Comment by cortfritz** [comment]
totally!  love it!

i'm 90% complete on observe but let's totally decouple these.

**Comment by cortfritz** [comment]
it has been tricky so far!  very fun.  key innovation was how to identify the stack items in the genserver.

File: lib/undo_manager.ex (hunk 4 above)
Comment (cortfritz) [reply lib/undo_manager.ex]: > It doesn't seem to reach here because it is checked by when.

i haven't addressed this yet, fyi

**Comment by cortfritz** [comment]
i wasn't explicit: my vote is merge now. :)
